### PR TITLE
fix(storage): enforce gRPC writer chunk retry deadlines by tying timers to server acks

### PR DESCRIPTION
**Pull request by vipnydav**
**Issue**
The `ChunkRetryDeadline` (default 32s) was not respected during transient network failures, trapping the gRPC writer in an infinite loop. This occurred due to three layered bugs:
1. **Premature Resets:** The timer was tied to local channel operations (`withCommandRetryDeadline`). Pushing bytes to a local channel is nearly instantaneous, so the timer was instantly cleared to `0` long before the actual network upload was attempted.
2. **Defeated Math:** When recovering a stream, the timer was unconditionally reset to `Now() + 32s`. Because the maximum backoff sleep is `30s`, fast-failing requests pushed the deadline infinitely into the future, making it mathematically impossible to expire.
3. **The `%w` Trap:** The timeout error was formatted using `%w`, wrapping the underlying `5xx` network error. The global retry loop unwrapped it, saw the transient network error, and blindly restarted the inner loop from scratch.

**Fixes**
* **Removed `withCommandRetryDeadline`:** Detached the timer entirely from local channel push operations.
* **Centralized the Stopwatch:** Initialized the timer at the top of the internal retry loop (`writeLoop`) only when establishing or recovering a stream, acting as a strict "time since last forward progress" watchdog.
* **Tied Resets to Server Progress:** Exclusively reset the timer inside `handleCompletion()`. This guarantees the timer only resets if the remote server physically acknowledges durably saving the bytes.
* **Changed `%w` to `%v`:** Updated the timeout error formatting so the global retry loop sees it as a fatal local failure and cleanly aborts instead of endlessly retrying.

### CODE REVIEW

File: storage/grpc_writer.go [hunk 1]
```diff
@@ -427,9 +423,12 @@ func (w *gRPCWriter) gatherFirstBuffer() error {
...
 	w.attempts++
+	if w.abandonRetriesTime.IsZero() {
+		w.abandonRetriesTime = time.Now().Add(w.chunkRetryDeadline)
+	}
 	// Return an error if we've been waiting for a single operation for too long.
-	if !w.abandonRetriesTime.IsZero() && time.Now().After(w.abandonRetriesTime) {
-		return fmt.Errorf("storage: retry deadline of %s reached after %v attempts; last error: %w", w.chunkRetryDeadline, w.attempts, w.lastErr)
+	if time.Now().After(w.abandonRetriesTime) {
+		return fmt.Errorf("storage: retry deadline of %s reached after %v attempts; last error: %v", w.chunkRetryDeadline, w.attempts, w.lastErr)
 	}
```
[the remark this answers is not in the document]
Comment (vipnydav) [reply storage/grpc_writer.go]: Go SDK's constructor already intercepts 0 and replaces it with 32s, so we do not need to add an `if w.chunkRetryDeadline > 0` check inside the `writeLoop`. The constructor has already guaranteed that it is safe!

File: storage/grpc_writer.go [hunk 2]
```diff
@@ -427,9 +423,14 @@ func (w *gRPCWriter) gatherFirstBuffer() error {
 
 func (w *gRPCWriter) writeLoop(ctx context.Context) error {
 	w.attempts++
-	// Return an error if we've been waiting for a single operation for too long.
-	if !w.abandonRetriesTime.IsZero() && time.Now().After(w.abandonRetriesTime) {
-		return fmt.Errorf("storage: retry deadline of %s reached after %v attempts; last error: %w", w.chunkRetryDeadline, w.attempts, w.lastErr)
+	if w.chunkRetryDeadline > 0 {
+		if w.abandonRetriesTime.IsZero() {
+			w.abandonRetriesTime = time.Now().Add(w.chunkRetryDeadline)
+		}
```
Comment (krishnamd-jkp) [inline storage/grpc_writer.go]: This does not define the start of the upload chunk, does it? This is a global place where the writer is starting but not the start of an individual chunk upload.

File: storage/grpc_writer.go [hunk 3]
```diff
@@ -363,19 +363,15 @@ func (w *gRPCWriter) handleCompletion(c gRPCBidiWriteCompletion) {
...
+
 	w.setSize(c.flushOffset)
 	w.progress(c.flushOffset)
 }
 
-func (w *gRPCWriter) withCommandRetryDeadline(f func() error) error {
-	w.abandonRetriesTime = time.Now().Add(w.chunkRetryDeadline)
-	err := f()
-	if err == nil {
-		w.abandonRetriesTime = time.Time{}
```
Comment (krishnamd-jkp) [inline storage/grpc_writer.go:382]: It is correct not to reset the timer here but the line 371 defines the correct start of the command. We might still have to keep this while also tying it with server responses.

Comment (vipnydav) [reply storage/grpc_writer.go:382]: Agreed. I've incorporated this at the very top of `writeLoop(ctx)`, right before we establish the new gRPC connection, to catch any chunks pending during a reconnect. And improved it by tying it to `w.isActive()`, ensuring the stopwatch accurately continues ticking for any pending chunks while we struggle to reconnect.

File: storage/grpc_writer.go (hunk 2 above)
Comment (vipnydav) [reply storage/grpc_writer.go]: I have updated the PR to gate the global check behind `isActive()`. This helps because it ensures that the timer only starts during a reconnect if there is actually data pending or in-flight (e.g., sitting in the channel waiting to dispatch). If the writer recovers but is completely idle, the timer safely stays paused, which perfectly prevents any false timeouts!

File: storage/grpc_writer.go [hunk 4]
```diff
@@ -509,6 +519,9 @@ Loop:
 					return errors.New("storage.Writer: unexpectedly closed w.writesChan")
 				}
 				w.currentCommand = cmd
+				if w.chunkRetryDeadline > 0 && w.abandonRetriesTime.IsZero() {
```
Comment (krishnamd-jkp) [inline storage/grpc_writer.go]: This is resetting the timer based on user commands which need not be the start of a chunk. There can be flush/close commands which has nothing to do with the start of a command. We might need to start the timer at a more accurate point. Like at the beginning of write command handle (in cases where we start sending the buffer where write size > buffer size)

File: storage/grpc_writer.go [hunk 5]
```diff
@@ -363,17 +368,17 @@ func (w *gRPCWriter) handleCompletion(c gRPCBidiWriteCompletion) {
...
 	w.setSize(c.flushOffset)
 	w.progress(c.flushOffset)
-}
 
-func (w *gRPCWriter) withCommandRetryDeadline(f func() error) error {
-	w.abandonRetriesTime = time.Now().Add(w.chunkRetryDeadline)
-	err := f()
-	if err == nil {
-		w.abandonRetriesTime = time.Time{}
+	if w.chunkRetryDeadline > 0 && w.isActive() {
```
Comment (krishnamd-jkp) [inline storage/grpc_writer.go:398]: Any reason we are doing this here as well? Because we are already pushing the timer in the write loop once we take the command off the channel.

**Comment by krishnamd-jkp** [comment]
Please add tests because this is a change in the core write logic, we will need to verify if these changes work fine in different cases. Like small writes which sit in the buffer for a while before another write and other corner case scenarios.

File: storage/grpc_writer.go (hunk 5 above)
Comment (vipnydav) [reply storage/grpc_writer.go:398]: They handle two different scenarios:

- The `writesChan` block starts the timer only when the writer is idle (`abandonRetriesTime.IsZero()`).
- The `handleCompletion` block restarts the timer after a chunk is successfully written. This is needed because if a new command arrived while the previous one was still in-flight, the `writesChan` block would have skipped starting a new timer. `handleCompletion` ensures the clock for the next chunk starts immediately after the previous one succeeds.

File: storage/grpc_writer.go (hunk 4 above)
Comment (vipnydav) [reply storage/grpc_writer.go]: I've moved the timer initialization directly to where the actual network requests are enqueued (`sendBufferToTarget`, the explicit `Flush` handler, and the` Close` tail request at the end of `writeLoop`) rather than when the user command is received. This ensures we strictly measure network/queueing time, completely ignoring user idle time.

File: storage/writer.go [hunk 6]
```diff
@@ -106,18 +106,23 @@ type Writer struct {
...
+	// error will be returned to the caller. This deadline measures the total
+	// time spent queuing, transmitting, and retrying a single chunk. For HTTP
+	// clients, a chunk is defined by the ChunkSize. For gRPC clients, data
+	// is streamed internally in 2 MiB quantums, and this deadline applies to
+	// each individual quantum.
+	//
+	// For HTTP clients, this is only applicable to files which are large
+	// enough to require a multi-chunk upload. For gRPC clients, this applies
+	// to all uploads. The default value is 32s. Users may want to pick a longer
+	// deadline if they are using larger values for ChunkSize or if they expect
```
Comment (krishnamd-jkp) [inline storage/writer.go]: This line is a bit confusing because for gRPC the chunksize doesn't matter because we are measuring the round trip time of quantum.

File: storage/grpc_writer_test.go [hunk 7]
```diff
@@ -412,3 +413,140 @@ func TestGRPCWriter_Deadlock(t *testing.T) {
...
+		}
+
+		// Attempt 1: Start the clock
+		_ = w.writeLoop(ctx)
+
+		// Sleep to consume more than half the deadline
+		time.Sleep(120 * time.Millisecond)
+
+		// Attempt 2: Clock should not be expired yet
+		err := w.writeLoop(ctx)
```
Comment (krishnamd-jkp) [inline storage/grpc_writer_test.go]: Check if the error is "transient network error"

File: storage/grpc_writer_test.go [hunk 8]
```diff
@@ -412,3 +413,140 @@ func TestGRPCWriter_Deadlock(t *testing.T) {
...
+			chunkSize:          100,
+			writesChan:         make(chan gRPCWriterCommand, 1),
+			setSize:            func(int64) {},
+			progress:           func(int64) {},
+		}
+
+		// Attempt 1: Start the clock
+		_ = w.writeLoop(ctx)
+
+		// Sleep to consume more than half the deadline
```
Comment (krishnamd-jkp) [inline storage/grpc_writer_test.go]: nit: Add periods at the end. In all the comments in this PR.

File: storage/grpc_writer_test.go [hunk 9]
```diff
@@ -412,3 +413,140 @@ func TestGRPCWriter_Deadlock(t *testing.T) {
...
+	// Immediately close completions to simulate stream failure.
+	// writeLoop will detect this instantly and return errResult.
+	close(cs.completions)
+}
+
+func (i *instantFailSender) err() error {
+	return i.errResult
+}
+
+func TestGRPCWriter_ChunkRetryDeadline(t *testing.T) {
```
Comment (krishnamd-jkp) [inline storage/grpc_writer_test.go]: Add a test to see how the writer behaves when there are a lot of small writes that sit in the buffer. The final write which fills the buffer should start off the timer. Check this scenario. And try to check if you can convert this test into a table driven format or see if it makes sense to split this test into multiple ones

Comment (vipnydav) [reply storage/grpc_writer_test.go]: Added the requested test and also divided the tests into multiple tests.

File: storage/writer.go (hunk 6 above)
Comment (vipnydav) [reply storage/writer.go]: Updated the comment to differentiate between http and gRPC.

**Review by krishnamd-jkp — APPROVED** [review]

**Comment by krishnamd-jkp** [comment]
Will unblock it after storage release.

File: storage/writer.go [hunk 10]
```diff
@@ -106,18 +106,26 @@ type Writer struct {
...
+	// For HTTP clients, a chunk is defined by the ChunkSize, so this deadline
+	// is only applicable to files large enough to require a multi-chunk upload.
+	// Users may also want to pick a longer deadline if they are using larger
+	// values for ChunkSize.
+	//
+	// For gRPC clients, data is streamed internally in 2 MiB quantums,
+	// and this deadline applies to each individual quantum. Therefore, it applies
+	// to all uploads, and the deadline does not need to scale with the ChunkSize.
+	//
+	// The default value is 32s. Users may want to pick a longer deadline if they
```
Comment (cpriti-os) [inline storage/writer.go:134]: Isn't 32s too much for 2MB quantum?

File: storage/grpc_writer_test.go [hunk 11]
```diff
@@ -412,3 +418,205 @@ func TestGRPCWriter_Deadlock(t *testing.T) {
...
+	ctx := context.Background()
+	deadline := 100 * time.Millisecond
+	sender := &instantFailSender{errResult: errors.New("transient network error")}
+	w := &gRPCWriter{
+		chunkRetryDeadline: deadline,
+		streamSender:       sender,
+		settings:           &settings{},
+		bufUnsentIdx:       0, // Makes isActive() == false.
+		bufFlushedIdx:      0,
+		buf:                make([]byte, 100),
```
Comment (cpriti-os) [inline storage/grpc_writer_test.go]: This is not an empty buffer, but we are pausing, could you check why this test case pass, it shouldn't ideally right?

**Review by cpriti-os — CHANGES_REQUESTED** [review]

File: storage/grpc_writer_test.go (hunk 11 above)
Comment (vipnydav) [reply storage/grpc_writer_test.go]: Yeah. It should not pass ideally. I had missed checking that if `w.buf` has data more than `w.writeQuantum` than it is active.

Added that check now.

File: storage/writer.go (hunk 10 above)
Comment (vipnydav) [reply storage/writer.go:134]: It may sound like a lot for a 2MB quantum, but with exponential backoff (1s, 2s, 4s, 8s...), this only translates to about 4-5 actual retries. I didn't change it since it has always been the standard for both the HTTP/JSON and gRPC clients. It's configurable, so with the new documentation, users can easily understand the behavior and override it if they need to.

File: storage/grpc_writer_test.go [hunk 12]
```diff
@@ -412,3 +418,205 @@ func TestGRPCWriter_Deadlock(t *testing.T) {
...
+	w := &gRPCWriter{
+		chunkRetryDeadline: deadline,
+		streamSender:       sender,
+		settings:           &settings{},
+		bufUnsentIdx:       100, // Makes isActive() == true.
+		bufFlushedIdx:      0,
+		buf:                make([]byte, 100),
+		sendableUnits:      1,
+		writeQuantum:       100,
+		chunkSize:          100,
```
Comment (cpriti-os) [inline storage/grpc_writer_test.go:476]: I can see that in all cases chunkSize = writeQuantum which isn't ideal, could you add more tests.
This test points out an edge case, please check
```
func TestGRPCWriter_ChunkRetryDeadline_Bug2(t *testing.T) {
	ctx := context.Background()
	deadline := 100 * time.Millisecond
	sender := &mockSender{errResult: errors.New("transient network error"), failOnData: false}
	
	w := &gRPCWriter{
		chunkRetryDeadline: deadline,
		streamSender:       sender,
		settings:           &settings{},
		bufUnsentIdx:       0,
		bufFlushedIdx:      -1,
		buf:                make([]byte, 0, 1000),
		sendableUnits:      1,
		writeQuantum:       100,
		chunkSize:          1000,
		writesChan:         make(chan gRPCWriterCommand, 3),
		setSize:            func(int64) {},
		progress:           func(int64) {},
		setObj:             func(*ObjectAttrs) {},
	}

	errCh := make(chan error, 1)
	go func() {
		errCh <- w.writeLoop(ctx)
	}()

	// Write 150 bytes.
	// writeQuantum is 100.
	// 100 bytes will be sent, 50 bytes will remain unsent in buf.
	done := make(chan struct{})
	w.writesChan <- &gRPCWriterCommandWrite{p: make([]byte, 150), done: done}
	<-done
	
	// Wait a bit to ensure completion for the 100 bytes is processed.
	time.Sleep(50 * time.Millisecond)
	
	// Check if timer was cleared.
	timerCh := make(chan time.Time)
	w.writesChan <- &checkTimerCmd{timerCh: timerCh}
	abandonRetriesTime := <-timerCh
	
	if !abandonRetriesTime.IsZero() {
		t.Errorf("BUG! Timer was NOT cleared! It is keeping the clock running. Value: %v", abandonRetriesTime)
	}

	// 2. Wait for the old timer to expire
	time.Sleep(150 * time.Millisecond)

	// 3. Now the connection "breaks" on the next write
	sender.mu.Lock()
	sender.failOnData = true
	sender.mu.Unlock()

	// Write 50 more bytes to reach the next quantum (100)
	done2 := make(chan struct{})
	w.writesChan <- &gRPCWriterCommandWrite{p: make([]byte, 50), done: done2}
	
	// writeLoop will return transient error
	err := <-errCh
	if err == nil || !strings.Contains(err.Error(), "transient network error") {
		t.Fatalf("expected transient network error, got: %v", err)
	}

	// 4. In a real scenario, `run()` will retry by calling `writeLoop` again.
	err = w.writeLoop(ctx)
	
	// If the timer wasn't cleared, it will instantly fail with "retry deadline reached"
	if err != nil && strings.Contains(err.Error(), "retry deadline") {
		t.Fatalf("BUG: Retry failed instantly due to stale timer! err: %v", err)
	}
}
```

**Review by cpriti-os — CHANGES_REQUESTED** [review]

File: storage/grpc_writer_test.go (hunk 12 above)
Comment (vipnydav) [reply storage/grpc_writer_test.go:476]: I've added handling in `isActive()` calculation for remaining unsent bytes so the deadline timer pauses properly when `idle < writeQuantum`. Also added few more tests for partial buffer and single shot retry edge cases.

File: storage/grpc_writer_test.go [hunk 13]
```diff
@@ -292,7 +295,11 @@ func (m *mockSender) connect(ctx context.Context, cs gRPCBufSenderChans, opts ..
 				}
 			}
 
-			if req.flush {
+			if m.failOnData && (req.flush || len(req.buf) > 0) {
```
Comment (cpriti-os) [inline storage/grpc_writer_test.go]: AI found a race issue by running the test with a race detector, please check 
```
==================
WARNING: DATA RACE
Write at 0x00c001a8a4a0 by goroutine 701:
  cloud.google.com/go/storage.TestGRPCWriter_ChunkRetryDeadline_PartialQuantumCloseStaleTimer()
      /tmp/gcs-test/storage/grpc_writer_test.go:865 +0x774

Previous read at 0x00c001a8a4a0 by goroutine 703:
  cloud.google.com/go/storage.(*mockSender).connect.func1()
      /tmp/gcs-test/storage/grpc_writer_test.go:298 +0x44f
==================
--- FAIL: TestGRPCWriter_ChunkRetryDeadline_PartialQuantumCloseStaleTimer (0.17s)
```

**Review by cpriti-os — CHANGES_REQUESTED** [review]

File: storage/grpc_writer_test.go (hunk 13 above)
Comment (vipnydav) [reply storage/grpc_writer_test.go]: Thanks for catching this!

The root cause was that in `mockSender.connect`, `m.failOnData` and `m.respondToAllData` were being read outside of `m.mu.Lock()`. Meanwhile, the test routines mutate `sender.failOnData` under `sender.mu.Lock()`, which led to the data race in the test.

I've updated `mockSender.connect` to safely copy `failOnData` and `respondToAllData` under `m.mu.Lock()` while processing each request. 

I've just pushed a commit with this fix. Let me know if everything looks good now!

**Review by cpriti-os — APPROVED** [review]
